File: backend/app/types/gl_journals.py
from pydantic import BaseModel
from typing import Optional
from datetime import date
from decimal import Decimal


class GLJournalLineCreate(BaseModel):
    account_id: int
    description: Optional[str] = None
    debit: Decimal = Decimal("0")
    credit: Decimal = Decimal("0")
    dim1_value_id: Optional[int] = None
    dim2_value_id: Optional[int] = None
    dim3_value_id: Optional[int] = None
    dim4_value_id: Optional[int] = None
    dim5_value_id: Optional[int] = None


class CreateGLJournalRequest(BaseModel):
    journal_date: date
    memo: Optional[str] = None
    lines: list[GLJournalLineCreate]


class UpdateGLJournalRequest(BaseModel):
    journal_date: Optional[date] = None
    memo: Optional[str] = None
    lines: Optional[list[GLJournalLineCreate]] = None

File: frontend/app/lib/types/gl_journals.ts
export interface GLJournalLine {
	id?: number;
	account_id: number | null;
	description: string | null;
	debit: number;
	credit: number;
	dim1_value_id: number | null;
	dim2_value_id: number | null;
	dim3_value_id: number | null;
	dim4_value_id: number | null;
	dim5_value_id: number | null;
}

export interface GLJournal {
	id: number;
	journal_date: string;
	reference: string;
	memo: string | null;
	status: "draft" | "posted" | "voided";
	workflow_status: "pending" | "approved" | "rejected" | null;
	total_debit?: number;
	created_at: string;
	posted_at: string | null;
	lines?: GLJournalLine[];
}

export interface CreateGLJournalRequest {
	journal_date: string;
	memo?: string;
	lines: Omit<GLJournalLine, "id">[];
}

export interface UpdateGLJournalRequest {
	journal_date?: string;
	memo?: string;
	lines?: Omit<GLJournalLine, "id">[];
}

export interface DimensionValues {
	dim1_value_id: number | null;
	dim2_value_id: number | null;
	dim3_value_id: number | null;
	dim4_value_id: number | null;
	dim5_value_id: number | null;
}
